operator/docs/enhancements/zone_aware_replication.md

```diff
@@ -0,0 +1,358 @@
+
+```yaml
+apiVersion: loki.grafana.com/v1beta1
+kind: LokiStack
+metadata:
+  name: lokistack-dev
+spec:
+  size: 1x.small
+  storage:
+    secret:
+      name: test
+      type: s3
+  storageClassName: gp2
+  replication:
+    factor: 3
+    topology:
+      enabled: true
+      zones:
+      - key: topology.kubernetes.io/zone
+        maxSkew: 1
+```
+	
+#### General constraints
+
+### Risks and Mitigations
+
+## Design Details
+The loki components can be divided in the Write Path (Distributor, Ingester) and the Read Path (Query fronted, Querier, Ingester) .
+The first step would be to make sure that all components respectively their replicas are configured to be spread over different zones. This can be done making using of the `PodTopologySpreadConstraint` feature in Kubernetes.
+Kubernetes makes a few assumptions about the structure of zones and regions:
+* regions and zones are hierarchical: zones are strict subsets of regions and no zone can be in 2 regions
+* zone names are unique across regions; for example region "africa-east-1" might be comprised of zones "africa-east-1a" and "africa-east-1b"
+
+Commonly used node labels to identify domains are 
+* topology.kubernetes.io/region
+* topology.kubernetes.io/zone
+* kubernetes.io/hostname
+
+The user needs to be aware of the node labels set to identify the different topology domains. Node read operations are generally made available to admin/developer users in OCP. This should be provided in the Lokistack CR topology key so that the `podTopologySpreadConstraint` can use this schedule the pods accordingly 
+
+These are some of the checks needed before the deployment happens. 
+* Check if the Topology key provided by the user matches a valid node topology domain label
+* Verify that the replication factor defined is less than the number of available zones
+If both these conditions are satisfied, deploy the lokistack components using the `PodTopologySpreadConstraint` definition in the deployments/statefulsets Pod template spec.
+
+Another important consideration is that each of the loki component pod needs the specific zone name to be set in their configuration. There is no easy way to implement this since the Kubernetes Downward-API does not support exposing node labels (See issue https://github.com/kubernetes/kubernetes/issues/40610)
+
+
+A couple of solutions to implement passing the node labels to the Loki pods are listed below:
+ 
+1. The plan is to introduce a Admission Mutating Webhook, that would watch the `pods/binding` sub-resource for each Loki pods. This webhook can update the pod annotations to add the topology key-value pair(s) when it is being scheduled on a node. This annotation (having the zone information) can then be used in the container as an environment variable using the Downward API and expanded in the Loki configuration per pod.
```

Comment:
I'm not sure why this approach uses a MutatingWebhook at all, as it is not mutating the resource the webhook is watching (the `Binding`) but the `Pod`. It should be possible to do this in a normal "resource watcher" on the Binding resource instead.

That change would bring this solution closer to the one listed below, so maybe these two solutions are actually the same.

Reply:
I agree with you but I am still concerned if the whole MutatingWebhook in the admission control is about patching the pod before the kubelet ever starts the container. A simple resource watcher would probably come in to late and a pod restart would be in need, at least this is my mental model about how it works. I might be wrong.

Reply:
Using the MutatingWebhook to change the pod definition "before it's written" is how I originally understood the "solution with MutatingWebhook", but the example provided does not seem to do that but just issues a "normal update". This is why I argued that _currently_ the two solutions are almost identical and can be reduced to one.

Either I am misunderstanding the code at hand or there's still room for "a real MutatingWebhook solution" that adds the annotation before the Pod configuration is written, so that no "empty startup" can happen.